File: jar-auth/src/main/java/com/beancounter/auth/JwtRoleConverter.java
package com.beancounter.auth;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.convert.converter.Converter;
import org.springframework.lang.NonNull;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
import org.springframework.stereotype.Component;

@Component
public final class JwtRoleConverter implements Converter<Jwt, AbstractAuthenticationToken> {
  private final JwtGrantedAuthoritiesConverter defaultGrantedAuthoritiesConverter
      = new JwtGrantedAuthoritiesConverter();
  private String resourceId;
  private String realmClaim;

  public JwtRoleConverter () {
    this ("realm_access", "roles");
  }
  public JwtRoleConverter(@Value("${auth.realm.claim:realm_access}") String realmClaim,
                          @Value("${auth.realm.roles:roles}") String resourceId) {
    this.realmClaim = realmClaim;
    this.resourceId = resourceId;
  }

  Collection<? extends GrantedAuthority> extractResourceRoles(@NonNull final Jwt jwt) {
    Map<String, Collection<String>> resourceAccess = jwt.getClaim(realmClaim);
    Collection<String> resourceRoles;
    if (resourceAccess != null && (resourceRoles = resourceAccess.get(resourceId)) != null) {
      return resourceRoles.stream()
          .map(role -> new SimpleGrantedAuthority("ROLE_" + role.toLowerCase()))
          .collect(Collectors.toSet());
    }
    return Collections.emptySet();
  }

  public Collection<GrantedAuthority> getAuthorities(@NonNull final Jwt jwt) {
    return Stream
        .concat(defaultGrantedAuthoritiesConverter.convert(jwt).stream(),
            extractResourceRoles(jwt).stream())
        .collect(Collectors.toSet());
  }

  @Override
  public AbstractAuthenticationToken convert(@NonNull final Jwt source) {
    Collection<GrantedAuthority> authorities = getAuthorities(source);
    return new JwtAuthenticationToken(source, authorities);
  }


}

File: jar-auth/src/main/kotlin/com/beancounter/auth/server/JwtRoleConverter.kt
package com.beancounter.auth.server

import org.springframework.beans.factory.annotation.Value
import org.springframework.core.convert.converter.Converter
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
import org.springframework.stereotype.Component
import java.util.Locale
import java.util.stream.Collectors
import java.util.stream.Stream

/**
 * Binds oAuth role mappings to Spring Security compatible.
 */
@Component
class JwtRoleConverter @JvmOverloads constructor(
    @Value("\${auth.realm.claim:realm_access}") private val realmClaim: String = "realm_access",
    @Value("\${auth.realm.roles:roles}") private val resourceId: String = "roles",
) : Converter<Jwt, AbstractAuthenticationToken> {

    private val defaultGrantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter()

    override fun convert(jwt: Jwt) = JwtAuthenticationToken(jwt, getAuthorities(jwt))

    fun extractResourceRoles(jwt: Jwt): Collection<GrantedAuthority> {
        val resourceAccess = jwt.getClaim<Map<String, Collection<String>>>(
            realmClaim
        )
        var resourceRoles: Collection<String> = mutableListOf()
        return if (resourceAccess != null && resourceAccess[resourceId]
            .also { resourceRoles = it!! }
            != null
        ) {
            resourceRoles.stream()
                .map { role: String
                    ->
                    SimpleGrantedAuthority("ROLE_" + role.lowercase(Locale.getDefault()))
                }
                .collect(Collectors.toSet())
        } else emptySet()
    }

    @Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
    fun getAuthorities(jwt: Jwt): Collection<GrantedAuthority> = Stream
        .concat(
            defaultGrantedAuthoritiesConverter.convert(jwt).stream(),
            extractResourceRoles(jwt).stream()
        )
        .collect(Collectors.toSet())
}
